### PR TITLE
Allow creation of new workspace above top most workspace

### PR DESCRIPTION
**Pull request by FluxTape**
Add the layout option allow-workspace-above-first to allow adding new workspaces above the top most workspace. When this option is enabled the first and last workspace on a monitor are always empty (compared to just the last one) allowing the movement of windows/columns above the first workspace.

### CODE REVIEW

File: niri-config/src/lib.rs [hunk 1]
```diff
@@ -429,6 +429,8 @@ pub struct Layout {
     pub center_focused_column: CenterFocusedColumn,
     #[knuffel(child)]
     pub always_center_single_column: bool,
+    #[knuffel(child)]
+    pub allow_workspace_above_first: bool,
```
Comment (YaLTeR) [inline niri-config/src/lib.rs:433]: "allow" feels weird here. Maybe `empty-workspace-above-first`? Or `add-empty-workspace-above-first`?

File: src/layout/monitor.rs [hunk 2]
```diff
@@ -105,6 +105,18 @@ impl<W: LayoutElement> Monitor<W> {
         }
     }
 
+    /// adds a workspace above the top most one. Also modifies self.active_workspace_idx
```
Comment (YaLTeR) [inline src/layout/monitor.rs:108]:
```suggestion
```

**Review by YaLTeR — CHANGES_REQUESTED** [review]
Please add checks into `verify_invariants()` that when this flag is active, there is an empty workspace above first, similarly to checks about there being an empty workspace below last. See https://github.com/YaLTeR/niri/blob/8d08782eba427a46607a478d088744d023a292c3/src/layout/mod.rs#L1819

Also run the randomized tests as described here: https://github.com/YaLTeR/niri/wiki/Developing-niri#running-tests

I expect you'll need to fix several more places that assume there's only an empty workspace in the end.

File: niri-config/src/lib.rs (hunk 1 above)
Comment (FluxTape) [reply niri-config/src/lib.rs:433]: I agree, I like `empty-workspace-above-first`

**Comment by FluxTape** [comment]
I've made the necessary changes to `verify_invariants()` and fixed any issues that popped up. Anything else that needs changing?

File: proptest-regressions/layout/mod.txt [hunk 3]
```diff
@@ -0,0 +1,8 @@
+# Seeds for failure cases proptest has generated in the past. It is
```
Comment (YaLTeR) [inline proptest-regressions/layout/mod.txt:1]: Don't push this file; instead, convert it into a regular unit test. As far as I understand, these seeds change when changing the Op enum (happens frequently).

File: src/layout/mod.rs [hunk 4]
```diff
@@ -1819,16 +1822,34 @@ impl<W: LayoutElement> Layout<W> {
                 monitor.workspaces.last().unwrap().columns.is_empty(),
                 "monitor must have an empty workspace in the end"
             );
+            if monitor.options.empty_workspace_above_first {
+                assert!(monitor.workspaces.last().unwrap().columns.is_empty(),
```
Comment (YaLTeR) [inline src/layout/mod.rs:1826]:
```suggestion
                assert!(monitor.workspaces.first().unwrap().columns.is_empty(),
```

File: src/layout/mod.rs [hunk 5]
```diff
@@ -1819,16 +1822,34 @@ impl<W: LayoutElement> Layout<W> {
...
+                "first workspace must be empty workspace when empty_workspace_above_first is set")
+            }
 
             assert!(
                 monitor.workspaces.last().unwrap().name.is_none(),
                 "monitor must have an unnamed workspace in the end"
```
Comment (YaLTeR) [inline src/layout/mod.rs:1832]: Needs a corresponding assert for `first()`

File: src/layout/mod.rs [hunk 6]
```diff
@@ -1819,16 +1822,34 @@ impl<W: LayoutElement> Layout<W> {
...
             if monitor.workspace_switch.is_none() {
-                for (idx, ws) in monitor.workspaces.iter().enumerate().rev().skip(1) {
+                for (idx, ws) in monitor
+                    .workspaces
+                    .iter()
+                    .enumerate()
+                    .skip(pre_skip)
+                    .rev()
+                    .skip(1)
+                // skip last
```
Comment (YaLTeR) [inline src/layout/mod.rs:1851]:
```suggestion
                    // skip last
                    .skip(1)
```

File: src/layout/mod.rs [hunk 7]
```diff
@@ -3950,6 +3971,27 @@ mod tests {
...
                     }
 
+                    first.apply(&mut layout);
+                    layout.verify_invariants();
+                    second.apply(&mut layout);
+                    layout.verify_invariants();
+                    third.apply(&mut layout);
+                    layout.verify_invariants();
+                }
+                // same test with empty-workspace-above-first set to true
```
Comment (YaLTeR) [inline src/layout/mod.rs:3981]: Not sure I want to add this since it doubles the test time (which is already the longest I believe). Let's let proptest catch these issues, together with manual test cases.

File: src/layout/mod.rs [hunk 8]
```diff
@@ -4315,12 +4357,18 @@ mod tests {
         };
 
         let mon = monitors.into_iter().next().unwrap();
+        let expected_workspace_offset = if mon.options.empty_workspace_above_first {
```
Comment (YaLTeR) [inline src/layout/mod.rs:4360]: Why this check? The test is done with default options.

Comment (FluxTape) [reply src/layout/mod.rs:4360]: Ups, when testing I temporarily changed the Default, this is a remnant of that. I'll refactor it to it's own test.

File: src/layout/mod.rs (hunk 7 above)
Comment (FluxTape) [reply src/layout/mod.rs]: ok removed it again

File: proptest-regressions/layout/mod.txt (hunk 3 above)
Comment (FluxTape) [reply proptest-regressions/layout/mod.txt]: removed in latest commit

File: src/layout/mod.rs (hunk 5 above)
Comment (FluxTape) [reply src/layout/mod.rs:1842]: should be fixed in latest commit

Comment (FluxTape) [reply src/layout/mod.rs:1842]: looks like there are still more issues, I'll investigate them later

Comment (FluxTape) [reply src/layout/mod.rs:2120]: Ok all regressions should be fixed now, can you take another look?

**Review by YaLTeR — CHANGES_REQUESTED** [review]
I'm so sorry, I forgot to submit the review several days ago..

**Comment by FluxTape** [comment]
Implemented all your suggestions, let me know if there are any more issues

**Comment by YaLTeR** [comment]
If you don't mind, I will postpone this to after the upcoming release. The logic has some remaining bugs, which are not difficult to fix, but they cannot get caught by the current testing system. I want to extend the testing system a bit to make sure they are caught and that there are no other similar bugs left.

**Comment by YaLTeR** [comment]
So that I don't forget, the current bugs are:
1. Interactive move uses add_tile() but assumes the workspace idx doesn't change, which it does if a workspace is added at the top. Possibly similar bugs elsewhere in the code. Not caught because the testing system needs to be able to advance animations to complete a workspace switch up.
2. Toggling the option at runtime doesn't add or remove the topmost workspace. Not caught because there are no tests that change the options.

**Comment by FluxTape** [comment]
> So that I don't forget, the current bugs are:
> 
>     1. Interactive move uses add_tile() but assumes the workspace idx doesn't change, which it does if a workspace is added at the top. Possibly similar bugs elsewhere in the code. Not caught because the testing system needs to be able to advance animations to complete a workspace switch up.
> 
>     2. Toggling the option at runtime doesn't add or remove the topmost workspace. Not caught because there are no tests that change the options.

To fix 2) it should be enough to add 
```rust
        if self.options.empty_workspace_above_first != options.empty_workspace_above_first
            && self.workspaces.len() > 1
        {
            if options.empty_workspace_above_first {
                self.add_workspace_top();
            } else {
                self.workspaces.remove(0);
                self.active_workspace_idx = self.active_workspace_idx.saturating_sub(1);
            }
        }
```
to `update_config(..)` in `monitor.rs`

**Comment by YaLTeR** [comment]
Working on advancing animations in the test system in #825.

**Comment by YaLTeR** [comment]
I've been running the randomized operations test a lot, and I've also implemented a proptest-state-machine version of the test that is a bit smarter in terms of generating valid sequences of operations in hopes that it will be more efficient at finding the issue 1 edge case. Unfortunately, I couldn't seem to get it to find it faster than the existing random operations test, and it adds quite a bit of code, so I kept it in a private branch for now. It did find one oversight in the AddWindowRightOf test implementation though which I just committed to main.

Both of these random tests manage to find the second output version of issue 1, but only after several minutes of runtime. Neither of the random tests has yet managed to find the single-output version of issue 1, which is slightly worrying. Makes me suspicious that there might be other problems that the current randomized tests have a hard time catching (probably due to the complexity and amount of operations).

Maybe it would be worth to also hook up cargo-fuzz with its coverage guidance. It uses a different Arbitrary trait though...

I'm not blocking this PR on this, just commenting what I've been spending time on. I'll next add a test for issue 2 and then the fix from your suggestion, and give it some more manual testing.

**Comment by FluxTape** [comment]
For 2) you can take a look at [this branch](https://github.com/FluxTape/niri/tree/workspace-above-v3) for inspiration, which helped me find and fix some issues.

**Comment by YaLTeR** [comment]
Thanks, I used that as a reference, but made a slightly simpler variation. Which immediately found an existing bug in update_config() :facepalm:

**Comment by YaLTeR** [comment]
With the config fix, one more problem that is not caught by tests: when enabling ewaf during a workspace switch animation or gesture, it gets glitched because its indices are now one off.

Need to check if it's ok to update the indices, or better to just cancel the gesture/animation.

Also, probably want to preserve the first empty workspace when disabling ewaf in case it was focused.

**Review by YaLTeR — APPROVED** [review]

**Comment by YaLTeR** [comment]
Note to self, remember to add it to the config wiki page

**Comment by YaLTeR** [comment]
I'm adding some ad-hoc conditions to the randomized tests, and I caught another "workspace index changed underneath" issue:
1. Start a workspace switch gesture on the first empty workspace.
2. Open a window.
3. The workspace index jumps up, but the index stored in the gesture remains unchanged.

I'll see if maybe simply stopping the gesture in `add_workspace_top()` is the best solution for now. Outside of this PR, maybe the workspace switch anim/gesture should store a workspace ID instead? Though, it won't magically fix all problems either. Need better tests for this.

**Comment by YaLTeR** [comment]
Another thing: need to make it so at startup with named workspaces, the named workspace is focused, rather than the first empty above named.

**Comment by YaLTeR** [comment]
Haven't found any more issues, so let's merge this. Thanks!
